[InteractionRegions] Add support for all non-uniform border-radius

etiennesegonzac opened the pull request:
#### b52ad6b19d1a80071f1aef489edf65ae1abcaed1
<pre>
[InteractionRegions] Add support for all non-uniform border-radius
<a href="https://bugs.webkit.org/show_bug.cgi?id=266491">https://bugs.webkit.org/show_bug.cgi?id=266491</a>
&lt;<a href="https://rdar.apple.com/118884255">rdar://118884255</a>&gt;

Reviewed by Tim Horton.

To support all `border-radius` configurations on InteractionRegions,
introduce an optional clip path.
Generate RoundedRect based Paths when the corner radii are not uniform.
Continue using masked corners when they can fully capture the shape.
Use `PathUtilities::pathWithShrinkWrappedRects` for elements with
multiple rects (like multi-line links).

* Source/WebCore/page/InteractionRegion.h:
Add an optional clip Path.
Rename borderRadius to cornerRadius to make it clearer that it works
with maskedCorners.
(WebCore::operator==):
Re-introduce a custom `==` operator. The compiler can&apos;t generate one now
that the struct contains a Path.
* Source/WebCore/page/InteractionRegion.cpp:
Remove the unused `PathUtilities.h` include.
(WebCore::interactionRegionForRenderedRegion):
Add an optional clip Path.
When the border radii are not uniform, generate a `clipPath` from
`RenderBox#borderRoundedRect()`.
(WebCore::operator&lt;&lt;):
Update the dump format for the new clipPath field.
Only dump the cornerRadius/masked corners when it is greater than 0.

* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::borderRoundedRect const):
Add a method we can use for to generate either clip paths or masked
corners.

* Source/WebCore/rendering/EventRegion.h:
Instead of accumulating the discovered rects for an element in a Region,
use a Vector&lt;FloatRect&gt;&gt; (to avoid rounding to an IntRect).
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::uniteInteractionRegions):
Accumulate the discovered rects for an element in a Vector instead of a
Region.
(WebCore::EventRegionContext::shouldConsolidateInteractionRegion):
Only consider consolidation with elements that have a single discovered
rect (can&apos;t consolidate to a multi-line link).
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):
Replace the shrink wrapping code with the great
`PathUtilities::pathsWithShrinkWrappedRects`.
This yields a single InteractionRegion with the proper rect and clip
path.
(WebCore::EventRegionContext::removeSuperfluousInteractionRegions):
Update the guard overlapping logic to iterate over all known interaction
rects (as opposed to checking once against each discovered Region).

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Add an optional clip Path.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintDebugOverlays):
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::InteractionRegionOverlay::activeRegion const):
(WebCore::InteractionRegionOverlay::drawRect):
borderRadius -&gt; cornerRadius

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::updateLayersForInteractionRegions):
Generate a CAShapeLayer from the clipPath when present, and use it to
mask the InteractionRegion layer.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _caLayerTreeAsText]):
Dump the layer tree of the content view instead of the whole WKWebView.
This makes the expectation files clearer and will help prevent churn
when unrelated part of the code change.
(dumpCALayer):
Add basic support for layer and corner masking in the dump.

* LayoutTests/interaction-region/border-radii.html:
* LayoutTests/interaction-region/border-radii-expected.txt:
This test already has non-uniform examples. Add borders and update
expectations.
* LayoutTests/interaction-region/guard-overlap.html:
* LayoutTests/interaction-region/guard-overlap-expected.txt:
Add a test with a guard overlapping a multi-line link.
* LayoutTests/interaction-region/layer-tree.html:
* LayoutTests/interaction-region/layer-tree-expected.txt:
Add various shape cases to the layer test to cover the UIProcess
changes.
* LayoutTests/interaction-region/layer-tree-shape-reset-expected.txt: Added.
* LayoutTests/interaction-region/layer-tree-shape-reset.html: Added.
Add a new test making sure all shaping properties on InteractionRegion
layers are properly reset if the style updates.
* LayoutTests/interaction-region/aria-roles-expected.txt:
* LayoutTests/interaction-region/button-in-link-expected.txt:
* LayoutTests/interaction-region/click-handler-dynamically-added-expected.txt:
* LayoutTests/interaction-region/click-handler-expected.txt:
* LayoutTests/interaction-region/click-handler-in-shadowed-layer-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/display-table-expected.txt:
* LayoutTests/interaction-region/event-region-overflow-expected.txt:
* LayoutTests/interaction-region/full-page-overlay-expected.txt:
* LayoutTests/interaction-region/hover-style-expected.txt:
* LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt:
* LayoutTests/interaction-region/ignore-catch-all-body-expected.txt:
* LayoutTests/interaction-region/inline-link-dark-background-expected.txt:
* LayoutTests/interaction-region/inline-link-expected.txt:
* LayoutTests/interaction-region/inline-link-in-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-in-layer-expected.txt:
* LayoutTests/interaction-region/inline-link-in-non-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-with-pointer-events-none-content-expected.txt:
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* LayoutTests/interaction-region/interaction-layers-culling-expected.txt:
* LayoutTests/interaction-region/interaction-layers-culling-layer-type-change-expected.txt:
* LayoutTests/interaction-region/interaction-region-size-limit-expected.txt:
* LayoutTests/interaction-region/labels-expected.txt:
* LayoutTests/interaction-region/nested-composited-text-painter-expected.txt:
* LayoutTests/interaction-region/overlap-expected.txt:
* LayoutTests/interaction-region/overlap-same-group-expected.txt:
* LayoutTests/interaction-region/overlay-expected.txt:
* LayoutTests/interaction-region/paused-video-regions-expected.txt:
* LayoutTests/interaction-region/position-only-update-expected.txt:
* LayoutTests/interaction-region/pseudo-element-expected.txt:
* LayoutTests/interaction-region/split-inline-link-expected.txt:
* LayoutTests/interaction-region/text-controls-expected.txt:
* LayoutTests/interaction-region/text-input-expected.txt:
* LayoutTests/interaction-region/tiny-regions-expected.txt:
* LayoutTests/interaction-region/wrapped-inline-link-expected.txt:
Update test expectations with clip paths and the new dump format.

Canonical link: <a href="https://commits.webkit.org/272377@main">https://commits.webkit.org/272377@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/539fdd6ae8e5d85a5a9016e190e34ae82221495f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31278 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9948 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32971 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33780 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28370 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32045 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12293 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7202 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28020 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31614 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8384 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27937 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7203 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7373 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27835 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35123 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28451 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28291 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33504 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7427 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5473 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31341 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9103 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/27630 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7383 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8126 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7951 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -244,92 +232,59 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(RenderObject& render
-            region.maskedCorners = maskedCorners;
+        Path path = PathUtilities::pathWithShrinkWrappedRects(discoveredRects, region.borderRadius);
+        region.rectInLayerCoordinates = layerBounds;
+        path.translate({ -layerBounds.location().x(), -layerBounds.location().y() });
```
```suggestion
        path.translate(-toFloatSize(layerBounds.location()));
```

pxlcoder commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -279,11 +263,16 @@ void updateLayersForInteractionRegions(RemoteLayerTreeNode& node)
+            if (region.clipPath) {
+                CAShapeLayer *mask = [regionLayer mask];
+                if (!mask) {
+                    mask = [[CAShapeLayer alloc] init];
```
```suggestion
                    mask = adoptNS([[CAShapeLayer alloc] init]).get();
```

etiennesegonzac replied:
was I leaking these? 🤦

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -281,18 +280,21 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
         auto borderRadii = renderBox->borderRadii();
         auto minRadius = borderRadii.minimumRadius();
         auto maxRadius = borderRadii.maximumRadius();
+        auto numberOfDifferentValues = HashSet<float>({ borderRadii.topLeft().minDimension(), borderRadii.topRight().minDimension(), borderRadii.bottomLeft().minDimension(), borderRadii.bottomRight().minDimension() }).size();
```
HashSet seems like wild overkill (especially given how hot this code is)

etiennesegonzac committed:
[InteractionRegions] Add support for all non-uniform border-radius
https://bugs.webkit.org/show_bug.cgi?id=266491
<rdar://118884255>

Reviewed by Tim Horton.

To support all `border-radius` configurations on InteractionRegions,
introduce an optional clip path.
Generate RoundedRect based Paths when the corner radii are not uniform.
Continue using masked corners when they can fully capture the shape.
Use `PathUtilities::pathWithShrinkWrappedRects` for elements with
multiple rects (like multi-line links).

* Source/WebCore/page/InteractionRegion.h:
Add an optional clip Path.
Rename borderRadius to cornerRadius to make it clearer that it works
with maskedCorners.
(WebCore::operator==):
Re-introduce a custom `==` operator. The compiler can't generate one now
that the struct contains a Path.
* Source/WebCore/page/InteractionRegion.cpp:
Remove the unused `PathUtilities.h` include.
(WebCore::interactionRegionForRenderedRegion):
Add an optional clip Path.
When the border radii are not uniform, generate a `clipPath` from
`RenderBox#borderRoundedRect()`.
(WebCore::operator<<):
Update the dump format for the new clipPath field.
Only dump the cornerRadius/masked corners when it is greater than 0.

* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::borderRoundedRect const):
Add a method we can use for to generate either clip paths or masked
corners.

* Source/WebCore/rendering/EventRegion.h:
Instead of accumulating the discovered rects for an element in a Region,
use a Vector<FloatRect>> (to avoid rounding to an IntRect).
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::uniteInteractionRegions):
Accumulate the discovered rects for an element in a Vector instead of a
Region.
(WebCore::EventRegionContext::shouldConsolidateInteractionRegion):
Only consider consolidation with elements that have a single discovered
rect (can't consolidate to a multi-line link).
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):
Replace the shrink wrapping code with the great
`PathUtilities::pathsWithShrinkWrappedRects`.
This yields a single InteractionRegion with the proper rect and clip
path.
(WebCore::EventRegionContext::removeSuperfluousInteractionRegions):
Update the guard overlapping logic to iterate over all known interaction
rects (as opposed to checking once against each discovered Region).

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Add an optional clip Path.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintDebugOverlays):
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::InteractionRegionOverlay::activeRegion const):
(WebCore::InteractionRegionOverlay::drawRect):
borderRadius -> cornerRadius

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::updateLayersForInteractionRegions):
Generate a CAShapeLayer from the clipPath when present, and use it to
mask the InteractionRegion layer.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _caLayerTreeAsText]):
Dump the layer tree of the content view instead of the whole WKWebView.
This makes the expectation files clearer and will help prevent churn
when unrelated part of the code change.
(dumpCALayer):
Add basic support for layer and corner masking in the dump.

* LayoutTests/interaction-region/border-radii.html:
* LayoutTests/interaction-region/border-radii-expected.txt:
This test already has non-uniform examples. Add borders and update
expectations.
* LayoutTests/interaction-region/guard-overlap.html:
* LayoutTests/interaction-region/guard-overlap-expected.txt:
Add a test with a guard overlapping a multi-line link.
* LayoutTests/interaction-region/layer-tree.html:
* LayoutTests/interaction-region/layer-tree-expected.txt:
Add various shape cases to the layer test to cover the UIProcess
changes.
* LayoutTests/interaction-region/layer-tree-shape-reset-expected.txt: Added.
* LayoutTests/interaction-region/layer-tree-shape-reset.html: Added.
Add a new test making sure all shaping properties on InteractionRegion
layers are properly reset if the style updates.
* LayoutTests/interaction-region/aria-roles-expected.txt:
* LayoutTests/interaction-region/button-in-link-expected.txt:
* LayoutTests/interaction-region/click-handler-dynamically-added-expected.txt:
* LayoutTests/interaction-region/click-handler-expected.txt:
* LayoutTests/interaction-region/click-handler-in-shadowed-layer-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/display-table-expected.txt:
* LayoutTests/interaction-region/event-region-overflow-expected.txt:
* LayoutTests/interaction-region/full-page-overlay-expected.txt:
* LayoutTests/interaction-region/hover-style-expected.txt:
* LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt:
* LayoutTests/interaction-region/ignore-catch-all-body-expected.txt:
* LayoutTests/interaction-region/inline-link-dark-background-expected.txt:
* LayoutTests/interaction-region/inline-link-expected.txt:
* LayoutTests/interaction-region/inline-link-in-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-in-layer-expected.txt:
* LayoutTests/interaction-region/inline-link-in-non-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-with-pointer-events-none-content-expected.txt:
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* LayoutTests/interaction-region/interaction-layers-culling-expected.txt:
* LayoutTests/interaction-region/interaction-layers-culling-layer-type-change-expected.txt:
* LayoutTests/interaction-region/interaction-region-size-limit-expected.txt:
* LayoutTests/interaction-region/labels-expected.txt:
* LayoutTests/interaction-region/nested-composited-text-painter-expected.txt:
* LayoutTests/interaction-region/overlap-expected.txt:
* LayoutTests/interaction-region/overlap-same-group-expected.txt:
* LayoutTests/interaction-region/overlay-expected.txt:
* LayoutTests/interaction-region/paused-video-regions-expected.txt:
* LayoutTests/interaction-region/position-only-update-expected.txt:
* LayoutTests/interaction-region/pseudo-element-expected.txt:
* LayoutTests/interaction-region/split-inline-link-expected.txt:
* LayoutTests/interaction-region/text-controls-expected.txt:
* LayoutTests/interaction-region/text-input-expected.txt:
* LayoutTests/interaction-region/tiny-regions-expected.txt:
* LayoutTests/interaction-region/wrapped-inline-link-expected.txt:
Update test expectations with clip paths and the new dump format.

Canonical link: https://commits.webkit.org/272377@main
